### PR TITLE
[LLVM][AArch64] Add assembly/disassembly for FTMOPA and BFTMOPA

### PR DESCRIPTION
**Pull request by SpencerAbson**
This patch adds assembly/disassembly for the following SME2p2 instructions (part of the 2024 AArch64 ISA update)

      -  BFTMOPA (widening) - FEAT_SME2p2
      -  BFTMOPA (non-widening)  - FEAT_SME2p2 & FEAT_SME_B16B16
      -  FTMOPA (4-way) - FEAT_SME2p2 & FEAT_SME_F8F32
      -  FTMOPA (2-way, 8-to-16) - FEAT_SME2p2 & FEAT_SME_F8F16
      -  FTMOPA (2-way, 16-to-32) - FEAT_SME2p2
      -  FTMOPA (non-widening, f16) - FEAT_SME2p2 & FEAT_SME_F16F16
      -  FTMOPA (non-widening, f32) - FEAT_SME2p2

- Add new ZPR_K register class and ZK register operand
- Introduce assembler extension tests for the new sme2p2 feature

In accordance with: https://developer.arm.com/documentation/ddi0602/latest/
Co-authored-by: Marian Lukac marian.lukac@arm.com

### CODE REVIEW

File: llvm/lib/Target/AArch64/Disassembler/AArch64Disassembler.cpp [hunk 1]
```diff
@@ -387,6 +389,23 @@ static DecodeStatus DecodeZPR2Mul2RegisterClass(MCInst &Inst, unsigned RegNo,
...
+static DecodeStatus DecodeZK(MCInst &Inst, unsigned RegNo, uint64_t Address,
+                             const MCDisassembler *Decoder) {
+  // RegNo <  4 => Reg is in Z20-Z23 (offset 20)
+  // RegNo >= 4 => Reg is in Z28-Z31 (offset 24)
+  unsigned Reg = (RegNo < 4) ? (RegNo + 20) : (RegNo + 24);
+  if (!(Reg >= 20 && Reg <= 23) && !(Reg >= 28 && Reg <= 31))
+    return Fail;
+
+  unsigned Register =
+      AArch64MCRegisterClasses[AArch64::ZPRRegClassID].getRegister(Reg);
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/Disassembler/AArch64Disassembler.cpp:404]: Can we use ZPR_KRegClassID instead of ZPRRegClassID?

**Review by CarolineConcatto — COMMENTED** [review]
Thank you Spencer for the patch.

**Review by jthackray — APPROVED** [review]
LGTM

File: llvm/lib/Target/AArch64/Disassembler/AArch64Disassembler.cpp [hunk 2]
```diff
@@ -387,6 +389,20 @@ static DecodeStatus DecodeZPR2Mul2RegisterClass(MCInst &Inst, unsigned RegNo,
   return Success;
 }
 
+// Zk Is the name of the control vector register Z20-Z23 or Z28-Z31, encoded in
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/Disassembler/AArch64Disassembler.cpp:392]: I believe we can remove the comments now that we are not using any math to compute the register number.

**Review by CarolineConcatto — APPROVED** [review]
Thank you for the patch Spencer.
I think we should remove the comments from static DecodeStatus DecodeZK as it does not makes sense anymore.

**Pull request merged**